### PR TITLE
Use a single stream for cuda graph pool

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97378
* #97440
* #97176
* #96775
* __->__ #97419
* #97171

Previously, we would use the same memory pool but not actually reuse the same memory. The peak memory showed good numbers, but real memory use was much higher because we had a bunch of unallocated segments that could not be reused. 

As stated in comments:

NB: cuda caching allocator will remember the stream a segment is allocated to
and only allocate that segment to the same stream. we need to use a single stream
for all allocations to the memory pool, otherwise the allocations to separate streams
will not be reused; separate recordings would have use the same memory pool, but not
the same memory.

Thanks to @zdevito for help debugging this.

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/cudagraph_trees.py [hunk 1]
```diff
@@ -323,26 +323,23 @@ def is_cuda_tensor(x):
...
-    with torch.device(device):
-        torch.cuda.synchronize()
-        stream = torch.cuda.Stream()
-        stream.wait_stream(torch.cuda.current_stream())
-        stream_context = torch.cuda.stream(stream)
-        stream_context.__enter__()
+    torch.cuda.synchronize()
+    stream.wait_stream(torch.cuda.current_stream())
+
+    with torch.cuda.stream(stream), torch.device(device):
```
Comment (wconstab) [inline torch/_inductor/cudagraph_trees.py:336]: what does stream(stream) do - create a new stream based on given stream?

Comment (eellison) [reply torch/_inductor/cudagraph_trees.py:336]: its a context manager for enabling the passed in stream

**Comment by ngimel** [comment]
Will all graphs be replayed on that single stream too?

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**